azure-iot-device/azure/iot/device/common/pipeline/pipeline_ops_base.py

```diff
@@ -162,3 +162,82 @@ def __init__(self, sas_token, callback=None):
         """
         super(SetSasToken, self).__init__(callback=callback)
         self.sas_token = sas_token
+
+
+class SendIotRequestAndWaitForResponse(PipelineOperation):
+    """
+    A PipelineOperation object which wraps the common operation of sending a request to iothub with a request_id ($rid)
+    value and waiting for a response with the same $rid value.  This convention is used by both Twin and Provisioning
+    features.
+
+    Even though this is an base operation, it will most likely be generated and also handled by more specifics stages
+    (such as IotHub or Mqtt stages).
+
+    The type of the request payload and the response payload is undefined at this level.  The type of the payload is defined
+    based on the type of request that is being executed.  If types need to be converted, that is the responsibility of
+    the stage which creates this operation, and also the stage which executes on the operation.
+
+    :ivar status_code: The status code returned by the response.  Any value under 300 is considered success.
+    :type status_code: int
+    :ivar response_body: The body of the response.
+    :type response_payload: Undefined
+    """
+
+    def __init__(self, request_type, verb, resource, request_body, callback=None):
+        """
+        Initializer for SendIotRequestAndWaitForResponse objects
+
+        :param str request_type: The type of request.  This is a string which is used by transport-specific stages to
+          generate the actual request.  For example, if request_type is "twin", then the iothub_mqtt stage will convert
+          the request into an MQTT publish with topic that begins with $iothub/twin
+        :param str verb: The verb for the request, such as "POST", "GET", etc.
+        :param str resource: The resource that the verb is acting on.  For twin request with verb "GET", this is most
```

Comment:
>resource [](start = 19, length = 8)

Like a URI?

Reply:
Yep, like the R in URI.  REST-over-MQTT FTW, baby!

---
In reply to: [288342591](https://github.com/Azure/azure-iot-sdk-python-preview/pull/96#discussion_r288342591) [](ancestors = 288342591)

Reply:
What I mean is, is the resource always provided in URI form? "resource" doesn't tell us much about what the parameter *is*, just what it represents. If it's a URI, lets call it a URI.

---
In reply to: [288677292](https://github.com/Azure/azure-iot-sdk-python-preview/pull/96#discussion_r288677292) [](ancestors = 288677292,288342591)

Reply:
It's not in URI form.  It can be _part_ of a URI, and, if this was REST instead of our bogus REST-over-MQTT, it would be used to build a URI (along with hostname and query parameters).  iothub steals the concept of a resource from REST, just like it steals the concept of a method.  For example, in the REST call to GET [https://api.github.com/repos/Azure/azure-iot-sdk-node/pulls/346](https://api.github.com/repos/Azure/azure-iot-sdk-node/pulls/346), the resource for this request (the entity that you're acting upon) is "/repos/Azure/azure-iot-sdk-node/pulls/346".

I think I can phrase this better in the comments by saying that it is the resource "in the REST API sense of the word" and I should add this same clause to the doc for the method param.

Does that address your concern, or am I still missing the point of your question?  


---
In reply to: [288742923](https://github.com/Azure/azure-iot-sdk-python-preview/pull/96#discussion_r288742923) [](ancestors = 288742923,288677292,288342591)

Reply:
No, that was the point I was getting at. How about resource_location?

---
In reply to: [288763007](https://github.com/Azure/azure-iot-sdk-python-preview/pull/96#discussion_r288763007) [](ancestors = 288763007,288742923,288677292,288342591)